pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -526,6 +529,32 @@ private void handleUpsert(ImmutableSegment immutableSegment) {
             _tableNameWithType));
     PartitionUpsertMetadataManager partitionUpsertMetadataManager =
         _tableUpsertMetadataManager.getOrCreatePartitionManager(partitionId);
+    return partitionUpsertMetadataManager;
+  }
+
+  /**
+   * Handle Upsert TTL during segment state change from CONSUMING to ONLINE.
+   *
+   * When a new segments commit, we will remove primary key indexes from heap that has timestamp < (eventTime - TTL).
+   * For all segments sealed before (eventTime-TTL), their validDocIds won't get updated. We do an one-time persistance.
+   */
+  protected void handleUpsertTTL(String segmentName, SegmentZKMetadata segmentZKMetadata,
+      UpsertTTLConfig upsertTTLConfig) {
+    PartitionUpsertMetadataManager partitionUpsertMetadataManager =
+        getPartitionUpsertMetadataManagerBySegment(segmentName);
+
+    assert segmentZKMetadata != null;
+    long endTimeInMillis = segmentZKMetadata.getEndTimeMs();
+    long expiredTimestamp = endTimeInMillis - upsertTTLConfig.getTtlInMs();
+
+    partitionUpsertMetadataManager.removeExpiredPrimaryKeys(expiredTimestamp);
+    partitionUpsertMetadataManager.persistSnapshotForStableSegment(expiredTimestamp);
```

Comment:
Should we use persistSnapshotForStableSegments because usually there are more than 1 stable segments per partition?

Reply:
gotcha, updated.